### ISSUE TITLE
Cannot narrow down type of another key of TypedDict

### ISSUE THREAD

**Issue by NCPlayz** [microsoft/pyright#1949]
**Describe the bug**

It seems that pyright does not narrow down types of other keys in a TypedDict when one of the keys are stored as a reference.

**To Reproduce**

(see below)

**Expected behavior**

The `if key == 'name':` statement should have an equivalent result to `if value['key'] == 'name':` (and `key` would naturally be narrowed down to `Literal['name']` 

**Screenshots or Code**

```python
from typing import List, Literal, TypedDict, Union


class StringValues(TypedDict):
    key: Literal['name']
    value: str

class IntValues(TypedDict):
    key: Literal['age']
    value: int

Values = Union[StringValues, IntValues]

values: List[Values] = [
    {'key': 'name', 'value': 'Monty'},
    {'key': 'age', 'value': 50},
]

for value in values:
    key = value['key']

    if key == 'name':
        reveal_type(key) # Type of "key" is "Literal['name']"
        reveal_type(value['key']) # Type of "value['key']" is "Literal['name', 'age']"
        reveal_type(value['value']) # Type of "value['value']" is "str | int"

    if value['key'] == 'name':
        reveal_type(key) # Type of "key" is "Literal['name', 'age']"
        reveal_type(value['key']) # Type of "value['key']" is "Literal['name']"!!!
        reveal_type(value['value']) # Type of "value['value']" is "str"!!!
```

**VS Code extension or command-line**
VS Code Extension (Pylance v2021.6.0)

**Comment by erictraut** [comment]
This is expected behavior. Static type checkers are not able to track associations between variables. In your example above, the code analysis engine doesn't track the fact that the value of `key` was accessed from the dictionary referenced by `value`. Tracking that type of association would be computationally intractable.
